### ISSUE TITLE
F821: false positive if variable is defined in a loop

### ISSUE THREAD

**Issue by JaRoSchm** [astral-sh/ruff#12451]
Hi,

this examples raises F821 in ruff 0.5.4 although it clearly is working:

```python
for i in range(3):
    if i == 0:
        x = 0
    else:
        x = y

    y = x + 2
```

Searching through the many issues where F821 is mentioned I couldn't find this issue yet.
Output of `ruff check --select F821 --preview`:

```
test.py:5:13: F821 Undefined name `y`
  |
3 |         x = 0
4 |     else:
5 |         x = y
  |             ^ F821
6 | 
7 |     y = x + 2
  |

Found 1 error.
```

**Comment by MichaReiser** [comment]
I think this is hard for ruff to get right at the moment without type inference because Ruff can't reason about that `i` starts with `0` and that the first branch is guaranteed to be taken because of that before ever taking the other branch.
